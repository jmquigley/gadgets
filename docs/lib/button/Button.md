<a name="Button"></a>

## Button
A typical button control widget.  This control only uses an icon and no text to represent the button.The icons are [Font Awesome](http://fontawesome.io/) strings.  That library is built into this module,so any font available in the current release of that library is available.#### Examples:```javascriptimport {Button} from 'gadgets';<Button iconName="cab" onClick={someFunction} />```#### Events- `onClick`` - when the button control is clicked by the user#### Styles- `ui-button` - top level style placed on `<i>` control that constructs the button.#### Properties- `classes: string[]` - an array of strings that represent custom CSS class names that will be appliedto the component.- `disabled: boolean` - if true then then the control is disabled.  Defaults to false.- iconName: string` - the name of the font awesome icon used with this button- `noripple: boolean` - turns off the ripple effect for a button.  On by default.- `style: object` - an object of key value/pairs that represent CSS style settings- `visible: boolean` - turns the display of this control on or off.  If true, then the component canbe seen.  The default is true.

**Kind**: global class  
