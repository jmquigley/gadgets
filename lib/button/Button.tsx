'use strict';

import {cloneDeep} from 'lodash';
import * as React from 'react';
import {baseClasses, BaseProps, getDefaultBaseProps} from '../shared';

const styles = require('./styles.css');

export interface ButtonProps extends BaseProps {
	iconName?: string;      // font awesome string
	style?: any;
}

export function getDefaultButtonProps(): ButtonProps {
	return cloneDeep(Object.assign(getDefaultBaseProps(), {
		iconName: 'bomb',
		style: {}
	}));
}

export const ButtonComponent = (props: ButtonProps) => (
    <i
        className={props.classes.join(' ')}
        onClick={props.onClick}
		aria-hidden="true"
	    disabled={props.disabled}
		style={props.style}
	>
	</i>
);

/**
 * A typical button control widget.  This control only uses an icon and no text to represent the button.
 * The icons are [Font Awesome](http://fontawesome.io/) strings.  That library is built into this module,
 * so any font available in the current release of that library is available.
 *
 * #### Examples:
 *
 * ```javascript
 * import {Button} from 'gadgets';
 * <Button iconName="cab" onClick={someFunction} />
 * ```
 *
 * #### Events
 * - `onClick`` - when the button control is clicked by the user
 *
 * #### Styles
 * - `ui-button` - top level style placed on `<i>` control that constructs the button.
 *
 * #### Properties
 * - `classes: string[]` - an array of strings that represent custom CSS class names that will be applied
 * to the component.
 * - `disabled: boolean` - if true then then the control is disabled.  Defaults to false.
 * - iconName: string` - the name of the font awesome icon used with this button
 * - `noripple: boolean` - turns off the ripple effect for a button.  On by default.
 * - `style: object` - an object of key value/pairs that represent CSS style settings
 * - `visible: boolean` - turns the display of this control on or off.  If true, then the component can
 * be seen.  The default is true.
 *
 */
export class Button extends React.Component<ButtonProps, undefined> {

    public static defaultProps: ButtonProps = getDefaultButtonProps();

    constructor(props: ButtonProps) {
		super(props);
	}

	private buildClasses = () => {
		let l: string[] = baseClasses(this.props);

		l.push(styles.button);
		l.push('fa');
		l.push(`fa-${this.props.iconName}`);
		l.push('ui-button');

		if (!this.props.noripple && !this.props.disabled) {
			l.push('ripple');
		}

		return l;
	}

	private handleClick = (e: Event) => {
		if (!this.props.disabled && this.props.visible && this.props.onClick != null) {
			this.props.onClick();
		}
		e.stopPropagation();
	}

	render() {
		return (
			<ButtonComponent
				{...this.props}
				onClick={this.handleClick}
				classes={this.buildClasses()}
			/>
		);
	}
}
