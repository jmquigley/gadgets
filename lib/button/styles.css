:local {
	.button {
		text-align: center;
		vertical-align: bottom;
		outline: none;
		width: 32px;
		background-color: inherit;
		line-height: inherit;
	}

	.button:not(.nohover):hover {
		background-color: lightgray;
	}

	.buttonDisabled {
		opacity: 0.5;
	}

	.buttonInvisible {
		display: none;
		width: 0;
	}
}
