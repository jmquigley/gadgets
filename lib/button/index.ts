import {IconSize} from '../icon';
import {Button, ButtonProps, getDefaultButtonProps} from './Button';

export {
	Button,
	ButtonProps,
	getDefaultButtonProps,
	IconSize
};
