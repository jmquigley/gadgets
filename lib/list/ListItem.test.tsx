"use strict";

import {mount, shallow} from "enzyme";
import * as React from "react";
import {waitPromise} from "util.wait";
import {Button} from "../button";
import {getDefaultListItemProps, ListItem} from "./index";

function validate(ctl: any) {
	expect(ctl).toBeDefined();
	expect(ctl.prop("disabled")).toBe(false);
	expect(ctl.prop("visible")).toBe(true);
	expect(ctl.find(".ui-listitem").length).toBe(10);
}

test("Test retrieval of ListItem props object", () => {
	const props = getDefaultListItemProps();

	expect(props).toBeDefined();
	expect(props).toMatchSnapshot();
});

test("Test the creation of a ListItem control with simple title", () => {
	const ctl = shallow(<ListItem title='test title' selected />);

	expect(ctl).toBeDefined();
	expect(ctl).toMatchSnapshot();
});

test("Test the creation of a ListItem control with left & right title", () => {
	const ctl = shallow(
		<ListItem title='test title left' widget='test title right' />
	);

	expect(ctl).toBeDefined();
	expect(ctl).toMatchSnapshot();
});

test("Test disabling of a ListItem control", () => {
	const click = jest.fn();
	const ctl = mount(
		<ListItem disabled={true} onClick={click} title='test title' />
	);

	expect(ctl).toBeDefined();
	expect(ctl).toMatchSnapshot();

	ctl.find(".ui-item")
		.first()
		.simulate("click");
	expect(click).not.toHaveBeenCalled();
});

test("Test making ListItem control invisible", () => {
	const click = jest.fn();
	const ctl = mount(<ListItem title='test title' visible={false} />);

	expect(ctl).toBeDefined();
	expect(ctl).toMatchSnapshot();

	ctl.find(".ui-item")
		.first()
		.simulate("click");
	expect(click).not.toHaveBeenCalled();
});

test("Test clicking of the left button on the ListItem control", () => {
	const click = jest.fn();
	const ctl = mount(
		<ListItem
			title='test title'
			leftButton={<Button iconName='bath' onClick={click} />}
		/>
	);

	validate(ctl);

	expect(ctl.prop("title")).toBe("test title");
	ctl.find(".ui-button")
		.first()
		.simulate("click");
	expect(click).toHaveBeenCalled();

	expect(ctl.find(".fa").length).toBe(3);
	expect(ctl.find(".fa-bath").length).toBe(3);
});

test("Test clicking of the right button on the ListItem control", () => {
	const click = jest.fn();
	const ctl = mount(
		<ListItem
			title='test title'
			rightButton={<Button iconName='bath' onClick={click} />}
		/>
	);

	validate(ctl);

	expect(ctl.prop("title")).toBe("test title");
	ctl.find(".ui-button")
		.last()
		.simulate("click");
	expect(click).toHaveBeenCalled();

	expect(ctl.find(".fa").length).toBe(3);
	expect(ctl.find(".fa-bath").length).toBe(3);
});

test("Test clicking of the title bar area of the ListItem", async () => {
	const click = jest.fn();
	const ctl = mount(<ListItem title='test title' onClick={click} />);

	validate(ctl);

	expect(ctl.prop("title")).toBe("test title");
	ctl.find(".ui-label")
		.first()
		.simulate("click");

	// This wait must occur during the test because there is a built in click
	// delay where the component checks if a double click is occurring.
	await waitPromise(1)
		.then(() => {
			expect(click).toHaveBeenCalled();
		})
		.catch((err: string) => {
			throw new Error(err);
		});
});

test("Test double click of the title bar area of the ListItem", () => {
	const click = jest.fn();
	const ctl = mount(<ListItem title='test title' onClick={click} />);
	const listItem = ctl.instance() as ListItem;

	validate(ctl);
	expect(listItem).toBeDefined();

	const titleCtl = ctl.find(".ui-label").first();
	expect(titleCtl).toBeDefined();

	expect(listItem.preventClick).toBe(false);
	expect(ctl.prop("title")).toBe("test title");
	titleCtl.simulate("doubleClick");
	expect(listItem.preventClick).toBe(true);
	expect(ctl.state("toggleRipple")).toBe(true);

	// The single click should be ignored because the double
	// click was used.
	expect(click).not.toHaveBeenCalled();

	// After the double click, simulate leaving focus and reseting
	// the control.
	titleCtl.simulate("blur");
	expect(listItem.preventClick).toBe(false);
	expect(ctl.state("toggleRipple")).toBe(false);
});
