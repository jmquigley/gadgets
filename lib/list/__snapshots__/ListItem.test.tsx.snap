// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test disabling of a ListItem control 1`] = `
<ListItem
  backgroundColor={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  contentEditable={false}
  controlled={true}
  direction="right"
  disabled={true}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  hiddenLeftButton={false}
  hiddenRightButton={false}
  href={
    Object {
      "selectHandler": [Function],
      "sizing": "normal",
    }
  }
  id={null}
  layout={5}
  left={null}
  leftButton={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noedit={false}
  nohover={false}
  notheme={false}
  notooltip={false}
  obj="ListItem"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[MockFunction]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onMouseOut={[Function]}
  onSelect={[Function]}
  onUpdate={[Function]}
  padding={null}
  rightButton={null}
  ripple={false}
  selected={false}
  sizing="normal"
  stacked={false}
  style={Object {}}
  testing={true}
  theme={null}
  title="test title"
  tooltip={null}
  top={null}
  useedit={false}
  visible={true}
  widget={null}
  width={null}
  xcss={null}
/>
`;

exports[`Test making ListItem control invisible 1`] = `
<ListItem
  backgroundColor={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  contentEditable={false}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  hiddenLeftButton={false}
  hiddenRightButton={false}
  href={
    Object {
      "selectHandler": [Function],
      "sizing": "normal",
    }
  }
  id={null}
  layout={5}
  left={null}
  leftButton={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noedit={false}
  nohover={false}
  notheme={false}
  notooltip={false}
  obj="ListItem"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onMouseOut={[Function]}
  onSelect={[Function]}
  onUpdate={[Function]}
  padding={null}
  rightButton={null}
  ripple={false}
  selected={false}
  sizing="normal"
  stacked={false}
  style={Object {}}
  testing={true}
  theme={null}
  title="test title"
  tooltip={null}
  top={null}
  useedit={false}
  visible={false}
  widget={null}
  width={null}
  xcss={null}
/>
`;

exports[`Test retrieval of ListItem props object 1`] = `
Object {
  "backgroundColor": null,
  "borderColor": null,
  "bottom": null,
  "children": null,
  "className": null,
  "color": null,
  "contentEditable": false,
  "controlled": true,
  "direction": "right",
  "disabled": false,
  "err": null,
  "errorMessage": null,
  "focus": false,
  "height": null,
  "hiddenLeftButton": false,
  "hiddenRightButton": false,
  "href": Object {
    "selectHandler": [Function],
    "sizing": "normal",
  },
  "id": null,
  "layout": 5,
  "left": null,
  "leftButton": null,
  "location": "none",
  "maxHeight": null,
  "maxWidth": null,
  "minHeight": null,
  "minWidth": null,
  "noedit": false,
  "nohover": false,
  "notheme": false,
  "notooltip": false,
  "obj": "ListItem",
  "onBlur": [Function],
  "onChange": [Function],
  "onClick": [Function],
  "onDoubleClick": [Function],
  "onFocus": [Function],
  "onKeyDown": [Function],
  "onKeyPress": [Function],
  "onMouseOut": [Function],
  "onSelect": [Function],
  "onUpdate": [Function],
  "padding": null,
  "rightButton": null,
  "ripple": false,
  "selected": false,
  "sizing": "normal",
  "stacked": false,
  "style": Object {},
  "testing": true,
  "theme": null,
  "title": "â€‹",
  "tooltip": null,
  "top": null,
  "useedit": false,
  "visible": true,
  "widget": null,
  "width": null,
  "xcss": null,
}
`;

exports[`Test retrieval of ListItem props object 2`] = `
Object {
  "children": null,
  "sizing": "normal",
  "style": Object {},
  "toggleRipple": true,
}
`;

exports[`Test the creation of a ListItem control with left & right title 1`] = `
<Wrapper
  backgroundColor={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  contentEditable={false}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  hiddenLeftButton={false}
  hiddenRightButton={false}
  href={
    Object {
      "selectHandler": [Function],
      "sizing": "normal",
    }
  }
  id={null}
  layout={5}
  left={null}
  leftButton={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noedit={false}
  nohover={false}
  notheme={false}
  notooltip={false}
  obj="ListItem"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onError={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onMouseOut={[Function]}
  onSelect={[Function]}
  onUpdate={[Function]}
  padding={null}
  reset={false}
  rightButton={null}
  ripple={false}
  selected={false}
  sizing="normal"
  stacked={false}
  style={Object {}}
  testing={true}
  theme={null}
  title="test title left"
  tooltip={null}
  top={null}
  useedit={false}
  visible={true}
  widget="test title right"
  width={null}
  xcss={null}
>
  <Item
    backgroundColor={null}
    borderColor={null}
    bottom={null}
    className="ui-listitem"
    color={null}
    contentEditable={false}
    controlled={true}
    direction="right"
    disabled={false}
    err={null}
    errorMessage={null}
    focus={false}
    height={null}
    hiddenLeftButton={false}
    hiddenRightButton={false}
    href={
      Object {
        "selectHandler": [Function],
        "sizing": "normal",
      }
    }
    id={null}
    layout={5}
    left={null}
    leftButton={null}
    location="none"
    maxHeight={null}
    maxWidth={null}
    minHeight={null}
    minWidth={null}
    noedit={false}
    nohover={false}
    notheme={false}
    notooltip={false}
    obj="ListItem"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onDoubleClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onMouseOut={[Function]}
    onSelect={[Function]}
    onUpdate={[Function]}
    padding={null}
    rightButton={null}
    ripple={true}
    selected={false}
    sizing="normal"
    stacked={false}
    style={Object {}}
    testing={true}
    theme={null}
    title="test title left"
    tooltip={null}
    top={null}
    useedit={false}
    visible={true}
    widget="test title right"
    width={null}
    xcss={null}
  />
</Wrapper>
`;

exports[`Test the creation of a ListItem control with simple title 1`] = `
<Wrapper
  backgroundColor={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  contentEditable={false}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  hiddenLeftButton={false}
  hiddenRightButton={false}
  href={
    Object {
      "selectHandler": [Function],
      "sizing": "normal",
    }
  }
  id={null}
  layout={5}
  left={null}
  leftButton={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noedit={false}
  nohover={false}
  notheme={false}
  notooltip={false}
  obj="ListItem"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onError={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onMouseOut={[Function]}
  onSelect={[Function]}
  onUpdate={[Function]}
  padding={null}
  reset={false}
  rightButton={null}
  ripple={false}
  selected={true}
  sizing="normal"
  stacked={false}
  style={Object {}}
  testing={true}
  theme={null}
  title="test title"
  tooltip={null}
  top={null}
  useedit={false}
  visible={true}
  widget={null}
  width={null}
  xcss={null}
>
  <Item
    backgroundColor={null}
    borderColor={null}
    bottom={null}
    className="ui-listitem ui-selected"
    color={null}
    contentEditable={false}
    controlled={true}
    direction="right"
    disabled={false}
    err={null}
    errorMessage={null}
    focus={false}
    height={null}
    hiddenLeftButton={false}
    hiddenRightButton={false}
    href={
      Object {
        "selectHandler": [Function],
        "sizing": "normal",
      }
    }
    id={null}
    layout={5}
    left={null}
    leftButton={null}
    location="none"
    maxHeight={null}
    maxWidth={null}
    minHeight={null}
    minWidth={null}
    noedit={false}
    nohover={false}
    notheme={false}
    notooltip={false}
    obj="ListItem"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onDoubleClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onMouseOut={[Function]}
    onSelect={[Function]}
    onUpdate={[Function]}
    padding={null}
    rightButton={null}
    ripple={true}
    selected={true}
    sizing="normal"
    stacked={false}
    style={Object {}}
    testing={true}
    theme={null}
    title="test title"
    tooltip={null}
    top={null}
    useedit={false}
    visible={true}
    widget={null}
    width={null}
    xcss={null}
  />
</Wrapper>
`;
